examples/pytorch/tgn/dataloading.py

```diff
@@ -0,0 +1,731 @@
+                                                    [neg_edge[1]],
+                                                    timestamp=ts)[0]
+            subg.ndata['timestamp'] = ts.repeat(subg.num_nodes())
+            batch_graphs.append(subg)
+        blocks = [dgl.batch(batch_graphs)]
+        input_nodes = torch.cat(nodes_id)
+        return input_nodes, pair_graph, neg_pair_graph, blocks
+
+    def collator(self, items):
+        """
+        The interface of collator, input items is edge id of the attached graph
+        """
+        result = super().collate(items)
+        # Copy the feature from parent graph
+        _pop_subgraph_storage(result[1], self.g)
+        _pop_subgraph_storage(result[2], self.g)
+        _pop_blocks_storage(result[-1], self.g_sampling)
+        return result
+
+
+class TemporalEdgeDataLoader(dgl.dataloading.EdgeDataLoader):
+    """ TemporalEdgeDataLoader is a iteratable object to generate blocks for temporal embedding
+    as well as pos and neg pair graph for memory update.
+
+    The batch generated will follow temporal order
+
+    Parameters
+    ----------
+    g : dgl.Heterograph
+        graph for batching the temporal edge id as well as generate negative subgraph
+
+    eids : torch.tensor() or numpy array
+        eids range which to be batched, it is useful to split training validation test dataset
+
+    block_sampler : dgl.dataloading.BlockSampler
+        temporal neighbor sampler which sample temporal and computationally depend blocks for computation
+
+    device : str
+        'cpu' means load dataset on cpu
+        'cuda' means load dataset on gpu
+
+    collator : dgl.dataloading.EdgeCollator
+        Merge input eid from pytorch dataloader to graph
+
+    Example
+    ----------
+    Please refers to examples/pytorch/tgn/train.py
+
+    """
+
+    def __init__(self, g, eids, block_sampler, device='cpu', collator=TemporalEdgeCollator, **kwargs):
+        collator_kwargs = {}
+        dataloader_kwargs = {}
+        for k, v in kwargs.items():
+            if k in self.collator_arglist:
+                collator_kwargs[k] = v
+            else:
+                dataloader_kwargs[k] = v
+        self.collator = collator(g, eids, block_sampler, **collator_kwargs)
+
+        assert not isinstance(g, dgl.distributed.DistGraph), \
+            'EdgeDataLoader does not support DistGraph for now. ' \
+            + 'Please use DistDataLoader directly.'
+        self.dataloader = torch.utils.data.DataLoader(
+            self.collator.dataset, collate_fn=self.collator.collate, **dataloader_kwargs)
+        self.device = device
+
+        # Precompute the CSR and CSC representations so each subprocess does not
+        # duplicate.
+        if dataloader_kwargs.get('num_workers', 0) > 0:
+            g.create_formats_()
+
+# ====== Fast Mode ======
+
+# Part of code comes from PyG library
+# https://github.com/rusty1s/pytorch_geometric
```

Comment:
Please given specific file name here. We just don't want to point to the whole PyG repo :)